### ISSUE TITLE
Examples should be modules - Monolithic Examples are bad for amber.

### ISSUE THREAD

**Issue by tomrake** [amber-smalltalk/amber-examples#69]
The amber website has had a non functional ProfStef page for several months, I have propose this pull request https://github.com/amber-smalltalk/amber-website/pull/19 to address this issue. In developing this PR I have come to the conclusion that the examples are poorly modularized and do not  promote easy reuse. The real roadblock had been that all of amber-examples need to be fixed  while the trysmalltalk/profstef code had been finished - all the other code was not used by the website but was holding the website back because of monolithic construction of amber-examples.

Examples that can be built in a modular fashion should be build in amber. Amber has benefited from the decomposition of the single monolith is was several years ago into the amber, amber-cli, amber-dev modules. 

Lets make all new amber examples modules that can be independently updated.

The problem of package construction is a key issue raised by @herby  in this post:
https://groups.google.com/forum/#!topic/amber-lang/jtm3ZXiNIH4 is at play in this case. We do not have tools to make it easy to create modular packages. We should create them.

**Comment by herby** [comment]
I'd like to understand where you see the problems, because IMO:

1. amber-example are not monolithic at all (as they are being update in last cca two months by Hannes Hirzel). they are completely decoupled. amber-examples is repo which is a collection of examples, but each one stands completely on its own, _nothing_ is common (or, should not be). So, question: where do you see the monolith?

2. The post you mention does not say at all we cannot do modularized amber. In fact, we can since 0.13.0, it's all there. What the post says that no one (except me and maybe a few courageous early adopters like sebastian concept or philippe back) knows how to do it / finds it too cumbersome, so the IDE should provide help with those tasks. So, if I should put a question here, it is: where do you see problems with modularization of amber code (there are definitely problems with user friendliness of it, but where do you see actual problems with modularization itself)?

Herby

Tom Rake wrote
:
> The amber website has had a non functional ProfStef page for several 
> months, I have propose this pull request 
> amber-smalltalk/amber-website#19 
> <https://github.com/amber-smalltalk/amber-website/pull/19> to address 
> this issue. In developing this PR I have come to the conclusion that 
> the examples are poorly modularized and do not promote easy reuse. The 
> real roadblock had been that all of amber-examples need to be fixed 
> while the trysmalltalk/profstef code had been finished - all the other 
> code was not used by the website but was holding the website back 
> because of monolithic construction of amber-examples.
>
> Examples that can be built in a modular fashion should be build in 
> amber. Amber has benefited from the decomposition of the single 
> monolith is was several years ago into the amber, amber-cli, amber-dev 
> modules.
>
> Lets make all new amber examples modules that can be independently 
> updated.
>
> The problem of package construction 
is a key issue raised by @herby

**Comment by hhzl** [comment]
See https://github.com/amber-smalltalk/amber-examples/issues/53#issuecomment-59219562

every subdirectory is independent for v013

Not much left for the examples to be finished.
Most of it is merged into branch gh-53.

ProcessingJS could be made a git subproject.

For v014 other options may be considered.

**Comment by tomrake** [comment]
Ease of use and reuse is key to my view. The website clearly envisioned reuse of the profstef code. That code clearly should be maintained in a single place and that place should be separately addressable by reference in the package system, bower.

Perhaps when I use modular I mean modular and package system addressable. A module should offer the ability to be used in the same way that the examples use amber via the bower.json.

I am not demanding that this has be my way. I am trying to understand how others think about this issue. It is just that I get to anber-examples which is a box and I really want to use one of the things that is in that box. As modules each component in the box could require additional dependencies and requiring to satisfy those could lead to bloat and possible package conflicts.

I am also looking at this from the point of view of composition of the amber-website from modules amber and profstef.

**Comment by hhzl** [comment]
Let's finish porting to v013 first.

Pull request to merge gh-53 into master: 
https://github.com/amber-smalltalk/amber-examples/pull/70

Suggestion: Then as a next step TrySmalltalk could be moved into a separate repository (like [grunt-init-amber](https://github.com/amber-smalltalk/grunt-init-amber)) and be converted into a library. Then people building different Amber web sites with instructions and tutorials may easily use it as a building block.

**Comment by herby** [comment]
Tom Rake wrote:
> Ease of use and reuse is key to my view. The website clearly envisioned
> reuse of the profstef code. That code clearly should be maintained in a
> single place and that place should be separately addressable by
> reference in the package system, bower.
>
> Perhaps when I use modular I mean modular and package system
> addressable. A module should offer the ability to be used in the same
> way that the examples use amber via the bower.json.

Ah. Yes, as it's one big repo, it's one package in bower. I wasn't 
thinking of examples as directly reusable, though. More like a starting 
points for development of similar project. But yes, if you want them to 
be directly addressable, that is lacking (you can always copy an example 
directory to your projects verbatim, include package amd names from it 
into deploy.js, and after `grunt deploy` and index.html reload you've 
got the classes accessible).

Herby

**Comment by hhzl** [comment]
@tomrake Could you please do #74

**Comment by mkroehnert** [comment]
I agree with @herby that most of the examples in this repository show some parts of Amber and should not be separated into many projects.
It would be very tiresome, if you would have to clone a new repo for each example.

For some things like TrySmalltalk or the presentations it does make sense.
But I don't think that for example the presentations would benefit from being installable with bower.
